### DISCUSSION TITLE
[源码改动求助]关于额外日志信息的添加

### DISCUSSION THREAD

**Discussion by RakuLomis** [MetaCubeX/mihomo#1511] [Q&A] [answered]
在历史版本以及当前版本中，日志所记载的信息如下(使用tun模式)：

`[TCP] 198.18.0.1: 65147(chrome.exe) --> ogs.google.com: 443 match DomainKeyword(google) using 节点选择[节点名称]`

即显示了协议类型、本地IP、本地端口、应用进程名称、请求域名、服务器端口。

在此基础上，我想增加一些额外的信息，以能够使内核协助我完成一些流量分析任务：
1. 针对特定进程A所打开的**监听端口号**
2. 针对进程A的数据所打开的**远程通信端口**
3. 进程A通信的**目标IP地址**

目前根据所阅读的部分源码，我推测日志信息的相关内容在tunnel.go的`logMetadata`函数中，并且结合metadata.go中定义的数据类型来看3中的**目标IP地址信息**应该是相对容易添加的。如有谬误，还请指正。

关于1和2，感觉不仅需要增加metadata.go中数据类型定义的相关内容，还需要对process.go，tunnel.go中进行方法的增加和调整，目前还没有特别明显的思路，感觉要改动的代码挺多。劳烦做过相关开发研究的前辈能够指点一下，需要修改哪些文件中的代码才有可能达到我想要的效果。

学艺不精，万分感谢！

**Comment by xishang0128** [comment]
直接拿metadata.xxx就行了
https://github.com/MetaCubeX/mihomo/blob/a08aa10630813485d7899e65ff59c3fae56f8364/constant/metadata.go#L131-L160

**Reply by RakuLomis** [reply]
感谢您的回复。我也尝试了使用Metadata来获取inbound和outbound的四个ip与端口。但是，实际上Metadata能够获得的似乎只有inbound RemoteIP/Port, inbound LocalIP/Port, outbound RemoteIP/Port，即outbound LocalIP/Port，出站的IP号（最主要的其实是端口号）无法获取。目前我的构思是，从tunnel.go中的`handleTCPConn`函数中的两个连接`Conn`和`RemoteConn`来分别获取inbound和outbound的IP与端口。

我在`handleTCPConn`中额外添加的代码如下：
```go
log.Infoln("inLoc: %s, inRemote: %s;", conn.LocalAddr().String(), conn.RemoteAddr().String()) 
log.Infoln("outLoc: %s, outRemote: %s;", remoteConn.LocalAddr().String(), remoteConn.RemoteAddr().String())
```
第一行可以正常运行并且将inbound的IP和端口输出到日志中，但是第二行对于outbound的代码却出现了一个很大的问题。

当我将内核build为exe后，并且使用clash verge来运行时，每当触发了代理规则，比如访问了谷歌，内核便会崩溃，此时无法访问任何网络，无论是系统代理或者是TUN模式都是如此。但是，当我在服务器上直接运行go项目时，我是用curl来访问谷歌、油管等网站进行测试时，却可以正常的执行下去。

百思不得其解，还望指教。再次感谢。

**Reply by RakuLomis** [reply]
更换UI为mihomo party后，可以间断性的运行。即非常的不稳定，会经常关闭内核，部分日志信息如下：

```log
time="2024-09-21T11:05:30.9260209+08:00" level=info msg="[TCP] 198.18.0.1:60487(chrome.exe) --> fonts.gstatic.com:443 match DomainSuffix(gstatic.com) using 🚀节点选择[美国02-IEPL-倍率1.0]"
time="2024-09-21T11:05:30.9260209+08:00" level=info msg="inLoc: 198.18.0.1:56850, inRemote: 198.18.0.2:10056;"
time="2024-09-21T11:05:30.9260209+08:00" level=info msg="outLoc: 192.168.5.2:60488, outRemote: 221.5.100.174:2062;"
time="2024-09-21T11:05:32.3668074+08:00" level=debug msg="[Metadata PreHandle] error: fake DNS record 198.18.0.208 missing"
time="2024-09-21T11:05:32.3668074+08:00" level=debug msg="[Metadata PreHandle] failed to sniff a domain for connection 198.18.0.1:60489 --> 198.18.0.208:8000, give up"
time="2024-09-21T11:05:32.4317631+08:00" level=debug msg="[DNS] hijack udp:198.18.0.2:53 from 198.18.0.1:60869"
time="2024-09-21T11:05:32.4548011+08:00" level=debug msg="[Rule] use default rules"
time="2024-09-21T11:05:32.4548011+08:00" level=info msg="[TCP] 198.18.0.1:60490 --> googleads.g.doubleclick.net:443 match DomainSuffix(doubleclick.net) using 🍃应用净化[REJECT]"
time="2024-09-21T11:05:32.4548011+08:00" level=info msg="inLoc: 198.18.0.1:56850, inRemote: 198.18.0.2:10058;"
[Manager]: Core closed, code: 2, signal: null
[Manager]: Try Restart Core
time="2024-09-21T11:05:32.9555082+08:00" level=info msg="Start initial configuration in progress"
time="2024-09-21T11:05:32.9752014+08:00" level=info msg="Geodata Loader mode: memconservative"
time="2024-09-21T11:05:32.9752014+08:00" level=info msg="Geosite Matcher implementation: succinct"
time="2024-09-21T11:05:33.0006951+08:00" level=info msg="Initial configuration complete, total time: 27ms"
time="2024-09-21T11:05:33.0006951+08:00" level=info msg="Sniffer is loaded and working"
```

**Reply by xishang0128** [reply]
https://github.com/MetaCubeX/mihomo/blob/a08aa10630813485d7899e65ff59c3fae56f8364/tunnel/tunnel.go#L552-L553

尝试在这行添加?

**Reply by RakuLomis** [reply]
感谢回复。我尝试了您的提议，并且也试了其他的位置。可惜的是结果依然没有改变。在通过规则访问网站后（大概几秒钟），内核就会关闭并且尝试重启。日志中的信息如下：
```log
time="2024-09-21T15:36:33.497249+08:00" level=info msg="[TCP] 198.18.0.1:52769 --> www.googleadservices.com:443 match DomainKeyword(adservice) using 🛑全球拦截[REJECT]"
time="2024-09-21T15:36:33.497249+08:00" level=info msg="inLoc: 198.18.0.1:52707, inRemote: 198.18.0.2:10023;"
[Manager]: Core closed, code: 2, signal: null
[Manager]: Try Restart Core
time="2024-09-21T15:36:33.9332519+08:00" level=info msg="Start initial configuration in progress"
```
猜想或许是“全球拦截”和“应用净化”对IP的处理不太一样？
此外，我将一次运行的完整文件附上。
[log-analysis-2.txt](https://github.com/user-attachments/files/17083309/log-analysis-2.txt)

**Reply by xishang0128** [reply, marked as answer]
才注意你reject了,那么remoteConn的两个Addr就是nil的,所以发生了panic

判断一下nil就行

**Reply by RakuLomis** [reply]
感谢！总算解决了。看来是在debug时并没有访问到触发reject规则的域名，还是我疏忽了。

受益匪浅，再次谢谢您！
